articles/user-guide/host-programs.md

```diff
@@ -0,0 +1,455 @@
+Details on what these metrics indicate can be found [here](xref:microsoft.quantum.machines.resources-estimator#metrics-reported).
+
+### Command line execution summary
+<br/>
+<img src="./images/command_line_diagram.png" alt="Q# program from command line" width="700">
+
+
+## Q# with host programs
+
+With our Q# file in hand, an alternative to calling an operation directly from the command line is to use a *host program* in another classical language. 
+Specifically, this can be done with either Python or a .NET language such as C# or F# (for the sake of brevity we will only detail C# here).
+A little more setup is required to enable the interoperability, but those details can be found in the [install guides](xref:microsoft.quantum.install).
+
+In a nutshell, the situation now includes a host program file (e.g. `*.py` or `*.cs`) in the same location as our Q# file.
+It's now the *host* program that gets run, and in the course of its execution it calls specific Q# operations from the Q# file.
+The core of the interoperability is based on the Q# compiler making the contents of the Q# file accessible to the host program so that they can be called.
+
+One of the main benefits of using a host program is that the classical data returned by the Q# program can then be further processed in the host language.
+This could consist of some advanced data processing (e.g. something that can't be performed internally in Q#), and then calling further Q# actions based on those results, or something as simple as plotting the Q# results.
+
+The general scheme is shown here, and we discuss the specific implementations for Python and C# below. 
+A sample using an F# host program can be found at the [.NET interoperability samples](https://github.com/microsoft/Quantum/tree/master/samples/interoperability/dotnet).
+
+<br/>
+<img src="./images/host_program_diagram.png" alt="Q# program from a host program" width="700">
+
+> [!NOTE]
+> The `@EntryPoint()` attribute used for Q# command line applications cannot be used with host programs.
+> An error will be raised if it is present in the Q# file being called by a host. 
+
+To work with different host programs, there are no changes required to a `*.qs` Q# file.
+The following host program implementations all work with the same Q# file:
+
+```qsharp
+namespace NamespaceName {
+    open Microsoft.Quantum.Intrinsic;     // for the H operation
+    open Microsoft.Quantum.Measurement;   // for MResetZ and MultiM
+	open Microsoft.Quantum.Canon;         // for ApplyToEach
+
+    operation MeasureSuperposition() : Result {
+        using (q = Qubit()) { 
+            H(q);
+            return MResetZ(q);
+        }
+    }
+
+    operation MeasureSuperpositionArray(n : Int) : Result[] {
+        mutable resultArray = new Result[n];
+        using (qubits = Qubit[n]) {
+            ApplyToEach( H , qubits);
+            set resultArray = MultiM(qubits);
```

Comment:
We've generally not encouraged MultiM compared to `M.Q.Arrays.ForEach(M, qubits)`. In this case, you could substitute `MResetZ` for `M` to avoid the mutable declaration on L255 and return directly from within the `using`.